src/dtls13.c

```diff
@@ -0,0 +1,2432 @@
+       acknowledgment */
+    if (Dtls13RtxMsgNeedsAck(ssl, hs))
+        ssl->dtls13Rtx.sendAcks = 1;
+}
+
+static int Dtls13ProcessBufferedMessages(WOLFSSL *ssl)
+{
+    DtlsMsg *msg = ssl->dtls_rx_msg_list;
+    word32 idx = 0;
+    int ret = 0;
+
+    WOLFSSL_ENTER("Dtls13ProcessBufferedMessages()");
+
+    while (msg != NULL) {
+        idx = 0;
+
+        /* message not in order */
+        if (ssl->keys.dtls_expected_peer_handshake_number != msg->seq)
+            break;
+
+        /* message not complete */
+        if (msg->fragSz != msg->sz)
+            break;
+
+        ret = DoTls13HandShakeMsgType(
+            ssl, msg->msg, &idx, msg->type, msg->sz, msg->sz);
+        if (ret != 0)
+            break;
+
+        Dtls13MsgProcessed(ssl, msg->type);
+
+        ssl->dtls_rx_msg_list = msg->next;
+        DtlsMsgDelete(msg, ssl->heap);
+        msg = ssl->dtls_rx_msg_list;
+        ssl->dtls_rx_msg_list_sz--;
+    }
+
+    WOLFSSL_LEAVE("dtls13_process_buffered_messages()", ret);
+
+    return ret;
+}
+
+static int Dtls13NextMessageComplete(WOLFSSL *ssl) {
+    return ssl->dtls_rx_msg_list != NULL &&
+           ssl->dtls_rx_msg_list->fragSz == ssl->dtls_rx_msg_list->sz &&
+           ssl->dtls_rx_msg_list->seq ==
+               ssl->keys.dtls_expected_peer_handshake_number;
+}
+
+static inline int FragIsInOutputBuffer(WOLFSSL *ssl, const byte *frag)
+{
+    const byte *OutputBuffer = ssl->buffers.outputBuffer.buffer;
+    word16 OutputBufferSize = ssl->buffers.outputBuffer.bufferSize;
+
+    return frag >= OutputBuffer && frag < OutputBuffer + OutputBufferSize;
+}
+
+static int Dtls13SendFragFromBuffer(WOLFSSL *ssl, byte *output, word16 length)
+{
+    byte *buf;
+    int ret;
+
+    if (FragIsInOutputBuffer(ssl, output))
+        return BAD_FUNC_ARG;
+
+    ret = CheckAvailableSize(ssl, length);
+    if (ret != 0)
+        return ret;
+
+    buf = ssl->buffers.outputBuffer.buffer + ssl->buffers.outputBuffer.length;
+
+    XMEMCPY(buf, output, length);
+
+    ssl->buffers.outputBuffer.length += length;
+
+    return SendBuffered(ssl);
+}
+
+static int Dtls13SendNow(WOLFSSL *ssl, enum HandShakeType handshakeType)
+{
+    if (!ssl->options.groupMessages)
+        return 1;
+
+    if (handshakeType == client_hello || handshakeType == hello_retry_request ||
+        handshakeType == finished || handshakeType == session_ticket ||
+        handshakeType == session_ticket || handshakeType == key_update)
+        return 1;
+
+    return 0;
+}
+
+/* Handshake header DTLS only fields are not inlcuded in the transcript hash */
+int Dtls13HashHandshake(WOLFSSL *ssl, const byte *output, word16 length)
+{
+    const byte *idx;
+    int ret;
+
+    if (length < DTLS_HANDSHAKE_HEADER_SZ)
+        return BAD_FUNC_ARG;
+
+    idx = output;
+
+    /* msg_type(1) + length (3) */
+    ret = HashRaw(ssl, idx, OPAQUE32_LEN);
+    if (ret != 0)
+        return ret;
+
+    idx += OPAQUE32_LEN;
+
```

Comment:
length -= OPAQUE32_LEN;
and below then passing only length in.
Makes it cleaner to read.